proposed/building-arch-specific-applications.md

```diff
@@ -0,0 +1,806 @@
+* **Architecture-specific, self-contained** -- Applications of this kind target and can only run on a single architecture, like Windows x64 or Linux ARM32. They carry a runtime and one set of dependent native libraries for a single target architecture. They include an executable launcher for the given architecture. All application, runtime and native binaries are contained in a single directory.
+
+### Proposed CLI argument and MSBuild property changes
+
+The following changes are proposed for .NET Core 3.0. `build` will be the specific verb discussed in the following text. The behavior is intended to be the same for `publish` unless specifically called out.
+
+#### Framework-dependent architecture-netural applications
+
+Maintain framework-dependent architecture-netural applications as the default for `build`.
+  
+* Produces an application with native dependencies for all supported architectures.
+* Remove the capability to produce EXEs for this type of build.
+* If EXEs are desired for this build type, consider for a later release. It's a hard scenario to get right due to filename conflicts.
+
+ASP.NET and console templates should accept this default.
+
+#### Framework-dependent architecture-specific applications
+
+Expose a new opt-in build type, called architecture-specific framework-dependent.
+
+* Produces an application with native dependencies and an EXE for a single specified architecture.
+* This build type can be selected via
+  * MSBuild: `<ArchitectureSpecific>true</ArchitectureSpecific>`
+  * CLI: `--architecture-specific` or `-a`
+* A RID must be specified (or it is an error). Any of the following options is legal to specify a RID:
+  * MSBuild: `<RuntimeIdentifier>RID</RuntimeIdentifier>`
+  * CLI: `--runtime RID`, or `-r RID`
+  * MSBuild: `<RuntimeIdentifier>$(NETCoreSdkRuntimeIdentifier)</RuntimeIdentifier>` (means: use RID of SDK; this already works)
+* With this build type, `--self-contained` is `false` by default, even when a RID is specified. `--self-contained` or `--self-contained true` must be specified to produce a self-contained build.
+
+WPF and Windows Forms templates should opt-in to this build type, with the following pattern:
+
+```xml
+<ArchitectureSpecific>true</ArchitectureSpecific>
+<RuntimeIdentifier>win-x86</RuntimeIdentifier>
+```
+
+The `win-x86` RID is proposed as the default to match the 32-bit default for .NET Framework projects. There is a perf benefit and it will be more compatible with existing .NET Framework projects that may have 32-bit only dependencies. It does mean that we need to install the Windows 32-bit .NET Core runtime by default (the 32-bit SDK is not needed). Visual Studio should make it straightforward to switch the default RID to `win-x64`.
```

Comment:
There is also the twitter thread on this (https://twitter.com/runfaster2000/status/1119401606645080064), but I'm replying here since 280 characters isn't enough 😄.

**TL;DR;** I would like to see some numbers showing that there is a perf benefit (for a range of application types) to staying in 32-bit mode by default, as this is not my expectation. I would also like to see memory usage numbers and for us to determine if that is because of actually higher memory usage or because of more liberal settings in the JIT/GC that are taking advantage of the increased address space.

I will always buy that you can see less memory usage by running 32-bit. Pointers are half the size and most other things remain the same, so 64-bit applications will naturally see some increased memory usage. As compared to a native app, this can be slightly worse on .NET since we have overhead of the various information the JIT and GC have to track; in addition to user types. However, I don't believe that in common applications this is particularly egregious. I also do not buy that there is a perf benefit to remaining on a 32-bit application.

Every architecture/platform has an ABI (Application Binary Interface) that defines things like how parameters are passed between methods, which CPU registers must be saved by the caller/callee method, default alignment of resources, etc. 32-bit has been around a lot longer and CPUs had a lot less resources when the ABI was defined. This means that you have more scenarios under which values need to be stored and retrieved from memory in order to preserve these limited CPU resources (namely CPU registers). It also means that, since not every CPU can be guaranteed to have these newer resources, sometimes tradeoffs have to happen when passing values around. An example of this is floating-point values being returned on the x87 FPU stack; even if modern apps never use it otherwise. It also means that operating on data that is larger than 32-bit can be more complicated and require more instructions. x64 CPUs on the other hand brought a new baseline to the table in terms of available CPU resources, instructions that were available, etc. One of the primary benefits is that it doubled the number of available integer and floating-point registers and made SSE/SSE2 instructions a baseline.

On 32-bit Windows, the default calling convention is `__cdecl` for regular C/C++ code and `__stdcall` for most Win32 APIs. Due to the limited number of registers, all values are passed by pushing them onto the stack (to memory). This means that you have a forced memory write and read for every parameter you pass (and memory access, even cached, is not "cheap"). Values are returned in register when they are less than 8-bytes and as a hidden pointer to memory otherwise. For value types, this commonly means a shadow copy happens (the value type is copied to a location in memory, and a pointer to that location is passed to the callee). For reference types, this means you have a double indirection to lookup the contents. For floating-point values, these are returned on the x87 FPU stack, which isn't otherwise used in modern applications and generally requires additional work to get it back into the SIMD registers that are used. There are also other various things about the 32-bit ABI that make it less than ideal. You only have three registers to save values in across calls, otherwise values you need to preserve also have to be spilled to memory so the callee method doesn't overwrite them. Also, anytime you need to work with 64-bit data (`System.Int64` or `System.UInt64`) you have to take up two registers to hold the data, you have to do more operations to compute the result, and you you have to work harder to allow atomic operations when multithreading is important.

On 64-bit Windows, the default calling convention is generally referred to as "fastcall" and it takes advantage of the additional CPU resources that were introduced. Now, rather than passing values by storing them to memory, it will pass the first four arguments in register (completely eliding the need for a memory access). Value types larger than 8-bytes still need to be shadow-copied and floating-point values are passed and returned directly in the SIMD registers rather than the x87 FPU stack. There are now 3 more integer registers and 10 more floating-point registers to save values in that the callee won't overwrite and the callee itself has 4 more integer registers in which to use, allowing it to spill less stuff to memory. Working with 64-bit data is also much simpler, as everything becomes a single operation and atomicity is easier to provide.

Overall, this means that even though you might be taking up slightly more memory in your app; common scenarios can involve less overall memory access. Memory access itself is expensive (compared to register access; and is cheap compared to disk IO, which is cheap compared to network IO), especially when you have a cache miss, so reducing the number of times you have to read/write memory can be important. An example of this is if you were to take something like: https://github.com/tannergooding/demoapp, which is a .NET Core 3.0 and WPF-based demo app that renders 3D models. If you run this app in 32-bit vs 64-bit mode you'll see that, for something like the `landscape` model, there is a ~21% increase in memory (going from `~150mb` to `~183mb`) but a 41% perf increase (going from `~50fps` to `~70fps` average); I wrote this app for the MVP summit to demo hardware intrinsics and tested on a 4k monitor and an AMD Ryzen 1800X machine.

Likewise, I believe Roslyn underwent various testing/benchmarking and it was found that there was no perf degredation from running the compiler in 32-bit vs 64-bit mode. I would expect that many other applications will see similar (either a perf increase or no perf degredation due to being run in 64-bit mode). Many native and UWP apps on my box are 64-bit by default and show similar results when compared to their 32-bit counterparts.

Reply:
The overall performance will be better on 64-bit for some apps and it will be better on 32-bit for other apps. It is probably close to a wash on average, depends on your benchmark mix. Performance is not going to help to decide this.

I think we should be looking at the other factors to decide the default:

- 64-bit is more secure, e.g. much better address space randomization, other new OS and hardware security features may be 64-bit only. 
- 64-bit is conceptually simpler. You do not have to think about whether 32-bit is enough for your app.
- The industry is on path to drop and deprecate 32-bit support.
- 32-bit is more compatible with existing .NET Framework default.

Looking at this list, I would think 64-bit would be better forward-looking default.

Reply:
If it's 32 bit, we will be explaining it forever. And that default assumes the perf difference is both measurable and permanent.

Reply:
It's worth noting that this is proposed as a template default, not a hidden magic choice in the build. I think that's great and it means there's no "forever" element to the choice here. We can always change templates if circumstances change.

Reply:
The explanation of "why is it on me to remember to make this explicit" would be the only forever element.

Reply:
> proposed as a template default

Does that mean that the following is no longer going to work?

1. Download 64-bit .NET Core SDK
2. dotnet new console
3. dotnet run

If the template defaults to 32-bit, is this going to fail with "you need to install 32-bit runtime" error?

Reply:
I have switched the default to 64-bit in the proposal based on this (and other) feedback.